### PR TITLE
[FLINK-5348] [core] Support custom field names for RowTypeInfo

### PR DESCRIPTION
**Pull request by wuchong**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by tonycox** [comment]
Hi @wuchong , in this case I think you should adapt `getFieldIndex`

**Comment by wuchong** [comment]
Hi @tonycox ,  the field names of RowTypeInfo are guaranteed to be unique. So I think `getFieldIndex ` is fine with it. Do you have special case which `getFieldIndex` give a wrong result ?

**Comment by tonycox** [comment]
@wuchong sorry, my fault, didn't spot duplicate field names checkers. What about cases, Is flink-table case-sensitive?

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java [hunk 1]
```diff
@@ -54,6 +76,152 @@ public RowTypeInfo(TypeInformation<?>... types) {
...
+		}
+
+		String field = matcher.group(1);
+
+		Matcher intFieldMatcher = PATTERN_INT_FIELD.matcher(field);
+		TypeInformation<X> fieldType = null;
+		if (intFieldMatcher.matches()) {
+			// field expression is an integer
+			int fieldIndex = Integer.valueOf(field);
+			if (fieldIndex > this.getArity()) {
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java]: `TupleTypeBase.getFieldAt()` does check for index bounds as well and throws an `IndexOutOfBoundsException`.
So we could simplify this a bit here.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java [hunk 2]
```diff
@@ -54,6 +76,152 @@ public RowTypeInfo(TypeInformation<?>... types) {
...
+			}
+		} else {
+			field = matcher.group(1);
+
+			Matcher intFieldMatcher = PATTERN_INT_FIELD.matcher(field);
+			TypeInformation<?> fieldType = null;
+			if (intFieldMatcher.matches()) {
+				// field expression is an integer
+				int fieldIndex = Integer.valueOf(field);
+				if (fieldIndex > this.getArity()) {
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java]: `TupleTypeBase.getFieldAt()` does check for index bounds as well and throws an `IndexOutOfBoundsException`.
So we could simplify this a bit here.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java [hunk 3]
```diff
@@ -54,6 +76,152 @@ public RowTypeInfo(TypeInformation<?>... types) {
...
+					throw new InvalidFieldReferenceException(
+						"Row field expression \"" + field + "\" out of bounds of " + this.toString() + ".");
+				}
+				for (int i = 0; i < fieldIndex; i++) {
+					offset += this.getTypeAt(i).getTotalFields();
+				}
+				fieldType = this.getTypeAt(fieldIndex);
+			} else {
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java]: We could use `getFieldIndex()` here to translate the field name into an index and use a common code path with the int field index to compute offset and fetch type.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java [hunk 4]
```diff
@@ -54,6 +76,152 @@ public RowTypeInfo(TypeInformation<?>... types) {
...
+		TypeInformation<X> fieldType = null;
+		if (intFieldMatcher.matches()) {
+			// field expression is an integer
+			int fieldIndex = Integer.valueOf(field);
+			if (fieldIndex > this.getArity()) {
+				throw new InvalidFieldReferenceException(
+					"Row field expression \"" + field + "\" out of bounds of " + this.toString() + ".");
+			}
+			fieldType = this.getTypeAt(fieldIndex);
+		} else {
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java]: translate `field` into an index with `getFieldIndex()` and use a common path fetch the type?

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java [hunk 5]
```diff
@@ -54,6 +76,152 @@ public RowTypeInfo(TypeInformation<?>... types) {
...
+
+		this.fieldNames = new String[fieldNames.size()];
+
+		for (int i = 0; i < fieldNames.size(); i++) {
+			this.fieldNames[i] = fieldNames.get(i);
+		}
+	}
+
+	@Override
+	public void getFlatFields(String fieldExpression, int offset, List<FlatFieldDescriptor> result) {
```
Comment (wuchong) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java]: It is almost ported from `CaseClassTypeInfo.getFlatFields()` except the field index. The `CaseClassTypeInfo` is 1-based, but `RowTypeInfo` is 0-based.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java [hunk 6]
```diff
@@ -54,6 +76,152 @@ public RowTypeInfo(TypeInformation<?>... types) {
...
+				} else {
+					throw new InvalidFieldReferenceException(
+						"Nested field expression \"" + tail + "\" not possible on atomic type " + fieldType + ".");
+				}
+			}
+		}
+	}
+
+	@Override
+	public <X> TypeInformation<X> getTypeAt(String fieldExpression) {
```
Comment (wuchong) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java]: It is almost ported from CaseClassTypeInfo.getFlatFields() except the field index. The CaseClassTypeInfo is 1-based, but RowTypeInfo is 0-based.

**Comment by wuchong** [comment]
Thank you @fhueske ,  I think your points are very good. And I changed my code according to your suggestions.

**Comment by fhueske** [comment]
Thanks for the update @wuchong!
PR is good to merge

**Comment by fhueske** [comment]
merging

**Pull request closed without merging**